src/main/java/org/trebol/jpa/services/predicates/command/PredicateCommand.java

```diff
@@ -0,0 +1,71 @@
+import org.trebol.jpa.entities.QProductCategory;
+import org.trebol.jpa.services.IProductCategoryTreeResolver;
+
+import java.util.List;
+import java.util.Map;
+
+public class PredicateCommand {
+  private static final String ID = "id";
+  private static final String CODE = "code";
+  private static final String NAME = "name";
+  private static final String NAME_LIKE = "nameLike";
+  private static final String PARENT_CODE = "parentCode";
+  private static final String PARENT_ID = "parentId";
+  private static final String ROOT_ID = "rootId";
+  private static final String ROOT_CODE = "rootCode";
+
+  private PredicateCommand() {}
+  private static final Map<String, Command> PREDICATE_MAP = Map.of(
+    ID, (stringValue, treeResolver, category, booleanBuilder) -> booleanBuilder.and(category.id.eq(Long.valueOf(stringValue))),
+    CODE, (stringValue, treeResolver, category, booleanBuilder) -> booleanBuilder.and(category.code.eq(stringValue)),
+    NAME, (stringValue, treeResolver, category, booleanBuilder) -> booleanBuilder.and(category.name.eq(stringValue)),
+    NAME_LIKE, (stringValue, treeResolver, category, booleanBuilder) -> booleanBuilder.and(category.name.likeIgnoreCase("%" + stringValue + "%")),
+    PARENT_CODE, (stringValue, treeResolver, category, booleanBuilder) -> {
+      if (StringUtils.isNotBlank(stringValue)) {
+        return booleanBuilder.and(category.parent.code.eq(stringValue));
+      }
+      return booleanBuilder;
+    },
+    PARENT_ID, (stringValue, treeResolver, category, booleanBuilder) -> {
+      if (StringUtils.isNotBlank(stringValue)) {
+        return booleanBuilder.and(category.parent.isNull());
+      } else {
+        return booleanBuilder.and(category.parent.id.eq(Long.valueOf(stringValue)));
+      }
+    },
+    ROOT_ID, (stringValue, treeResolver, category, booleanBuilder) -> {
+      if (StringUtils.isNotBlank(stringValue)) {
+        List<Long> branchParentIds = treeResolver.getBranchIdsFromRootId(Long.valueOf(stringValue));
+        return booleanBuilder.and(category.parent.id.in(branchParentIds));
+      }
+      return booleanBuilder;
+    },
+    ROOT_CODE, (stringValue, treeResolver, category, booleanBuilder) -> {
+      if (StringUtils.isNotBlank(stringValue)) {
+        List<Long> branchParentIds = treeResolver.getBranchIdsFromRootCode(stringValue);
+        if (!CollectionUtils.isEmpty(branchParentIds)) {
+          return booleanBuilder.and(category.parent.id.in(branchParentIds));
+        }
+      }
+      return booleanBuilder;
+    }
+  );
+
+
+  public static Predicate getPredicate(String param, String stringValue, IProductCategoryTreeResolver treeResolver, QProductCategory category, BooleanBuilder booleanBuilder) {
+    Command command = PREDICATE_MAP.get(param);
+
+    if (command == null) {
+      throw new IllegalArgumentException("Invalid param type: "
+        + param);
+    }
+
+    return command.getPredicate(stringValue, treeResolver, category, booleanBuilder);
+  }
+}
```

Comment:
This is a big refactor, you're basically setting up a new boilerplate for `IPredicateJpaService`s. 
But you are making this coming together nicely. 

It would be great to actually write the other services like this; it takes advantage of lambda functions and hash tables, instead of using a `switch` loop. I understand this should help performance but I'd have to try measuring it to fully grasp its potential.